fix: Switch from Prisma to TypeORM

mrv777 opened the pull request.

eandersson commented on migrations/1710000000000-InitialMigration.ts:
```diff
@@ -0,0 +1,159 @@
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`
+            CREATE TABLE "pool_stats" (
```
This is currently broken as pool_stats -> PoolStats etc.

mrv777 replied:
Just pushed a fix for this.  Sorry about that

eandersson replied:
Works now. Thanks!

eandersson commented on lib/db.ts:
```diff
@@ -1,18 +1,58 @@
+  database: process.env.DB_NAME || 'postgres',
+  entities: [PoolStats, User, UserStats, Worker, WorkerStats],
+  logging: process.env.NODE_ENV === 'development',
+  ssl: {
```
It would be nice to be able to configure ssl.

mrv777 replied:
I attempted to add this with https://github.com/mrv777/ckstats/pull/6/commits/d293a8560d7f3fae5d9600ea76a5cf54f2044315 but I didn't test it

eandersson commented on lib/db_migration.ts:
```diff
@@ -0,0 +1,35 @@
+  entities: [PoolStats, User, UserStats, Worker, WorkerStats],
+  migrations: ['migrations/*.ts'],
+  logging: process.env.NODE_ENV === 'development',
+  ssl: {
```
SSL stuff looks good but also needed here

eandersson replied:
Actually maybe not
> 230.2 19:7  Error: Replace `·process.env.DB_SSL·===·'true'` with `⏎····process.env.DB_SSL·===·'true'⏎·····`  prettier/prettier
> 230.2 20:1  Error: Insert `······`  prettier/prettier
> 230.2 21:3  Error: Replace `}` with `······}⏎·····`  prettier/prettier

mrv777 replied:
Forgot I had that prettier checks.  I'll push a fix there for the formatting, sorry about

eandersson replied:
This SSL call also needs to be configurable

eandersson commented:
Looking good. Only remaining issue besides ssl for me is that seed crashes and I don't understand the stack trace lol.

Example payload
```
{
    "runtime": 791013,
    "lastupdate": 1736704373,
    "Users": 2,
    "Workers": 9,
    "Idle": 2,
    "Disconnected": 3,
    "hashrate1m": "25.2T",
    "hashrate5m": "26.8T",
    "hashrate15m": "28.1T",
    "hashrate1hr": "29.6T",
    "hashrate6hr": "29.7T",
    "hashrate1d": "29.6T",
    "hashrate7d": "29.4T",
    "diff": 0.17,
    "accepted": 181899142101,
    "rejected": 110240305,
    "bestshare": 2440231045496,
    "SPS1m": 0.179,
    "SPS5m": 0.191,
    "SPS15m": 0.2,
    "SPS1h": 0.21
}
```

```
diff --git a/scripts/seed.ts b/scripts/seed.ts
index 216eff6..6e778b5 100644
--- a/scripts/seed.ts
+++ b/scripts/seed.ts
@@ -6,10 +6,11 @@ async function fetchPoolStats() {
   console.log('Fetching pool stats...');
   const apiUrl = process.env.API_URL || 'https://solo.ckpool.org';
   const response = await fetch(`${apiUrl}/pool/pool.status`);
-  const data = await response.text();
-  const jsonLines = data.split('\n').filter(Boolean);
-  const parsedData = jsonLines.reduce((acc, line) => ({ ...acc, ...JSON.parse(line) }), {});
-  return parsedData;
+  //const data = await response.text();
+  //const jsonLines = data.split('\n').filter(Boolean);
+  //const parsedData = jsonLines.reduce((acc, line) => ({ ...acc, ...JSON.parse(line) }), {});
+  //return pasedData;
+  return await response.json();
 }
```

```
131.7 > ckstats@0.3.0 migration:run /ckstats
131.7 > ts-node -P tsconfig.scripts.json ./scripts/migration.ts
131.7
149.4 Running migrations...
155.4
155.4 No pending migrations to run
155.4 Migration process completed
157.2
157.2 > ckstats@0.3.0 seed /ckstats
157.2 > ts-node -P tsconfig.scripts.json scripts/seed.ts
157.2
172.8 /ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:859
172.8     return new TSError(diagnosticText, diagnosticCodes, diagnostics);
172.8            ^
172.8 TSError: ⨯ Unable to compile TypeScript:
172.8 scripts/seed.ts(37,25): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(38,23): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(39,25): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(40,22): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(41,30): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(42,42): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(43,42): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(44,43): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(45,43): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(46,43): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(47,42): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(48,42): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(49,13): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(50,17): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(51,17): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(52,18): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(53,14): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(54,14): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(55,15): error TS18046: 'stats' is of type 'unknown'.
172.8 scripts/seed.ts(56,14): error TS18046: 'stats' is of type 'unknown'.
172.8
172.8     at createTSError (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:859:12)
172.8     at reportTSError (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:863:19)
172.8     at getOutput (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1077:36)
172.8     at Object.compile (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1433:41)
172.8     at Module.m._compile (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1617:30)
172.8     at node:internal/modules/cjs/loader:1699:10
172.8     at Object.require.extensions.<computed> [as .ts] (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1621:12)
172.8     at Module.load (node:internal/modules/cjs/loader:1313:32)
172.8     at Function._load (node:internal/modules/cjs/loader:1123:12)
172.8     at TracingChannel.traceSync (node:diagnostics_channel:322:14) {
172.8   diagnosticCodes: [
172.8     18046, 18046, 18046, 18046,
172.8     18046, 18046, 18046, 18046,
172.8     18046, 18046, 18046, 18046,
172.8     18046, 18046, 18046, 18046,
172.8     18046, 18046, 18046, 18046
172.8   ]
172.8 }
173.0  ELIFECYCLE  Command failed with exit code 1.
```

mrv777 commented:
> Looking good. Only remaining issue besides ssl for me is that seed crashes and I don't understand the stack trace lol.
> 
> Example payload
> 
> ```
> {
>     "runtime": 791013,
>     "lastupdate": 1736704373,
>     "Users": 2,
>     "Workers": 9,
>     "Idle": 2,
>     "Disconnected": 3,
>     "hashrate1m": "25.2T",
>     "hashrate5m": "26.8T",
>     "hashrate15m": "28.1T",
>     "hashrate1hr": "29.6T",
>     "hashrate6hr": "29.7T",
>     "hashrate1d": "29.6T",
>     "hashrate7d": "29.4T",
>     "diff": 0.17,
>     "accepted": 181899142101,
>     "rejected": 110240305,
>     "bestshare": 2440231045496,
>     "SPS1m": 0.179,
>     "SPS5m": 0.191,
>     "SPS15m": 0.2,
>     "SPS1h": 0.21
> }
> ```
> 
> ```
> diff --git a/scripts/seed.ts b/scripts/seed.ts
> index 216eff6..6e778b5 100644
> --- a/scripts/seed.ts
> +++ b/scripts/seed.ts
> @@ -6,10 +6,11 @@ async function fetchPoolStats() {
>    console.log('Fetching pool stats...');
>    const apiUrl = process.env.API_URL || 'https://solo.ckpool.org';
>    const response = await fetch(`${apiUrl}/pool/pool.status`);
> -  const data = await response.text();
> -  const jsonLines = data.split('\n').filter(Boolean);
> -  const parsedData = jsonLines.reduce((acc, line) => ({ ...acc, ...JSON.parse(line) }), {});
> -  return parsedData;
> +  //const data = await response.text();
> +  //const jsonLines = data.split('\n').filter(Boolean);
> +  //const parsedData = jsonLines.reduce((acc, line) => ({ ...acc, ...JSON.parse(line) }), {});
> +  //return pasedData;
> +  return await response.json();
>  }
> ```
> 
> ```
> 131.7 > ckstats@0.3.0 migration:run /ckstats
> 131.7 > ts-node -P tsconfig.scripts.json ./scripts/migration.ts
> 131.7
> 149.4 Running migrations...
> 155.4
> 155.4 No pending migrations to run
> 155.4 Migration process completed
> 157.2
> 157.2 > ckstats@0.3.0 seed /ckstats
> 157.2 > ts-node -P tsconfig.scripts.json scripts/seed.ts
> 157.2
> 172.8 /ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:859
> 172.8     return new TSError(diagnosticText, diagnosticCodes, diagnostics);
> 172.8            ^
> 172.8 TSError: ⨯ Unable to compile TypeScript:
> 172.8 scripts/seed.ts(37,25): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(38,23): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(39,25): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(40,22): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(41,30): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(42,42): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(43,42): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(44,43): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(45,43): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(46,43): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(47,42): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(48,42): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(49,13): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(50,17): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(51,17): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(52,18): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(53,14): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(54,14): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(55,15): error TS18046: 'stats' is of type 'unknown'.
> 172.8 scripts/seed.ts(56,14): error TS18046: 'stats' is of type 'unknown'.
> 172.8
> 172.8     at createTSError (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:859:12)
> 172.8     at reportTSError (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:863:19)
> 172.8     at getOutput (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1077:36)
> 172.8     at Object.compile (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1433:41)
> 172.8     at Module.m._compile (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1617:30)
> 172.8     at node:internal/modules/cjs/loader:1699:10
> 172.8     at Object.require.extensions.<computed> [as .ts] (/ckstats/node_modules/.pnpm/ts-node@10.9.2_@types+node@22.7.4_typescript@5.6.2/node_modules/ts-node/src/index.ts:1621:12)
> 172.8     at Module.load (node:internal/modules/cjs/loader:1313:32)
> 172.8     at Function._load (node:internal/modules/cjs/loader:1123:12)
> 172.8     at TracingChannel.traceSync (node:diagnostics_channel:322:14) {
> 172.8   diagnosticCodes: [
> 172.8     18046, 18046, 18046, 18046,
> 172.8     18046, 18046, 18046, 18046,
> 172.8     18046, 18046, 18046, 18046,
> 172.8     18046, 18046, 18046, 18046,
> 172.8     18046, 18046, 18046, 18046
> 172.8   ]
> 172.8 }
> 173.0  ELIFECYCLE  Command failed with exit code 1.
> ```

Looks like an annoying typescript error.  I added a type to see if that fixes it for you

eandersson commented:
It is unfortunately still failing with typescript errors. Is this due to the node version I am using or something along those lines?
